design/1.0.md

```diff
@@ -0,0 +1,91 @@
+I recently opened a Kubernetes Enhancement Proposal to introduce `go-flow-levee` during testing as a defense against accidental credential logging.
+See [KEP-1933](https://github.com/kubernetes/enhancements/pull/1936) details.
+Such integration demands a stable API and well-defined scope.
+
+I propose the following.
+Portions may already be implemented, but are included here as a "big picture."
+
+# 1.0 Targets
+
+## Basic taint propagation detection
+
+Given a spec defining sources and sinks (below), diagnostics are reported for
+any direct call to a sink that includes a source.  Additionally, we detect any
+taint propagation from a source to a sink within the scope of a single function.
+
+Analysis does not explicitly track across the use of reflection.
+
+## Analysis Configuration
+
+Configuration will consist of the following key components:
+
+* Identification of sources, sinks, and sanitizers
+* Identification of analysis scope, e.g. to skip third-party or testing packages.
+* Specify diagnostics to be ignored, e.g false-positives
+
+### Identification of sources, sinks, and sanitizers
+
+Much of our work is directly with `ssa.Value`s and `ssa.Instruction`s.
+Whether those values or instructions represent one element or another,
+ the configuration to identify these members should be as uniform as possible.
+In that light, specification should follow:
+
+#### ValueSpecifier
+A ValueSpecifier marks an `ssa.Value`, e.g., for identification as a source or as the safe result of a sanitizer.
+Values should be specifiable via any or all of:
+
+* Specifier name: For identification during reporting
+* Type / Field: Package path and type name, optional field name
+* Field tags
+* Scope: local variable, free variable, or global
+* Context: within a specific package or function call.  See CallSpecifier below.
+* Is a reference
+* Is "reference-like," e.g., slices, maps, chans.
+* Const value (e.g., `"PASSWORD"`)
+* Argument position (for use below)
+
+#### CallSpecifier
+A CallSpecifier marks an `*ssa.Call`, e.g., as a sink or sanitizer.
+Calls should be specifiable via any or all of:
+
+* Specifier name: For identification during reporting
+* Symbol: package path, function name, optional receiver name.
```

Comment:
Do we want to handle the following cases:
1. `func`s stored in `struct` fields, e.g.
```go
type MySinker struct {
  sinkFunc func(args ...interface{})
}
```

2. `func`s passed as function parameters:
```go
func callSink(arg interface{}, sink func(...interface{})) {
  sink(arg)
}
```
In theory, since `func`s are first class in Go, the actual range of cases we could potentially handle is much wider, e.g. a user could put functions in a `map[string]func()`, store functions in local variables, etc.

There are some cases we already handle, e.g.:
```go
func Test(s core.Source) {
  f := core.Sink
  f(s) // a diagnostic is produced here
}
```

But it is quite easy to throw our analysis off in this case:
```go
func Test(s core.Source) {
  var f func(...interface{})
  if true {
    f = core.Sink
  }
  f(s) // no diagnostic
}
```

My opinion: I think we should handle `func`s stored in `struct` fields, because that seems to be a fairly common pattern. The other cases seem more outlandish and seem like they would require some kind of cross-function or even whole-program analysis, so I think for a 1.0 they are out of scope.

Reply:
> My opinion: I think we should handle funcs stored in struct fields, because that seems to be a fairly common pattern. The other cases seem more outlandish and seem like they would require some kind of cross-function or even whole-program analysis, so I think for a 1.0 they are out of scope.

We currently rely on the static dispatch in our call matching.  In effect, we cover only case (a) of the four kinds of Call described [in the docs](https://godoc.org/golang.org/x/tools/go/ssa#CallCommon).

I think we should make a reasonable effort to cover all four cases.  (b) seems achievable in our current framework, and (c) seems a bit out-of-scope.  Dynamic dispatch in case (d) is probably a whole can of worms.  We could rely on `pointer.PointsTo` to get a set of possible functions, but that predicates whole-program analysis.

I don't think the fact that a given `func` is stored in a struct field will make the problem any simpler.  I'd be inclined to put this at a "post-1.0 but preferably 1.1" sort of target.

Reply:
Rereading this, I think well-defined specifiers catch both of these cases.  It's just a difference of a sink being identified by a `ValueSpecifier` rather than a `CallSpecifier`.  Pseudo-config in yaml for readability and ignoring implementation details:

```yaml
Sinks:
  ValueSpecifiers:
    - name: "MySinker matcher"
      symbol: "path/to/my/pkg.MySinker.sinkFunc"
    - name: "log wrapper"
      context:
       # This is a CallSpecifier
        symbol: "path/to/my/pkg.callSink"
      argPosition: 2
```